### ISSUE TITLE
Issue 18c: GCP Vertex AI training service

### ISSUE THREAD

**Issue by bmander** [bmander/thecityandthebike#63]
## Parent Issue
Part of #18 - Create image classification system

## Branch
`issue-18-c-training-service`

## Dependencies
- Issue 18a (#61, schema) must be merged first
- Issue 18h (#68, model architecture) must be merged first — provides the training script
- Can be built in parallel with issue 18b (#62, labels API)

## Scope
Large

## Description

Implement the ML service layer that orchestrates Vertex AI Custom Training jobs and manages scale-to-zero prediction endpoints, plus the training management API endpoints. This wraps the custom training script from #68 in a Vertex AI Custom Training Job.

### Approach: Vertex AI Custom Training (not AutoML)

Custom training gives us full control over the model architecture (defined in #68):
- TensorFlow + transfer learning (EfficientNet-B0 or similar)
- Version-controlled training code
- Tunable hyperparameters
- Lower per-node-hour cost than AutoML
- **Scale-to-zero endpoint** (`min_replica_count=0` via v1beta1 API) — no billing when idle

### New Python Dependency

```
google-cloud-aiplatform>=1.134.0
```

### New Configuration

Add to `api/app/config.py` Settings:

```python
ML_BACKEND: str = "mock"           # "vertex_ai" or "mock"
GCP_PROJECT: Optional[str] = None  # e.g., "tcatb-prod"
GCP_REGION: str = "us-central1"
```

### ML Service Module (`api/app/services/ml_service.py`)

**Training pipeline** (runs as FastAPI BackgroundTask):

```
Admin triggers POST /training/start
→ Validate sufficient labeled data (min ~10/class)
→ Create TrainingRun row (status=pending)
→ Return immediately with run_id
→ BackgroundTask:
  → Query labeled images from classification_labels + fender_submissions
  → Generate CSV manifest file:
      training,gs://bucket/images/abc.jpg,lime_left
      validation,gs://bucket/images/def.jpg,bird_right
  → Upload CSV to gs://bucket/ml/training-data/{run_id}.csv
  → Submit Vertex AI CustomTrainingJob using training script from #68
  → Poll job.state until complete
  → On success: deploy model with scale-to-zero, update TrainingRun + ClassificationModel
  → On failure: update TrainingRun with error_message
```

**Custom Training Job submission:**
```python
job = aiplatform.CustomTrainingJob(
    display_name=f"tcatb-fender-training-{run_id}",
    script_path="ml/train.py",
    container_uri="us-docker.pkg.dev/vertex-ai/training/tf-gpu.2-14.py310:latest",
    model_serving_container_image_uri="us-docker.pkg.dev/vertex-ai/prediction/tf2-gpu.2-14:latest",
)

model = job.run(
    model_display_name=f"tcatb-fender-model-{run_id}",
    args=["--epochs", "50", "--learning_rate", "1e-4", "--base_model", "EfficientNetB0",
          "--data_csv", f"gs://bucket/ml/training-data/{run_id}.csv"],
    replica_count=1,
    machine_type="n1-standard-4",
    accelerator_type="NVIDIA_TESLA_T4",
    accelerator_count=1,
)
```

**Model deployment** (after training completes):
```python
endpoint = model.deploy(
    deployed_model_display_name=f"tcatb-fender-{run_id}",
    machine_type="n1-standard-2",
    min_replica_count=0,   # scale to zero when idle — no billing
    max_replica_count=1,
)
# Store endpoint resource name in ClassificationModel row
```

**Prediction service:**
```python
def predict_vertex_ai(endpoint_resource_name, image_bytes) -> Prediction:
    endpoint = aiplatform.Endpoint(endpoint_resource_name)
    b64_content = base64.b64encode(image_bytes).decode("utf-8")
    response = endpoint.predict(instances=[{"content": b64_content}])
    # First call after idle may take ~30-60s (cold start)

def predict_mock(image_bytes) -> Prediction:
    # Return random prediction for dev/testing
```

**Mock service** (`ML_BACKEND=mock`, the default):
- Mock training: waits a few seconds, marks as complete
- Mock prediction: returns random class + confidence
- Enables full development without GCP access

### Training Router (`/training`)

| Method | Path | Auth | Description |
|--------|------|------|-------------|
| POST | `/training/start` | Admin | Trigger new training run |
| GET | `/training/status` | Auth | Current training status (is_training + current_run) |
| GET | `/training/runs` | Auth | List all training runs |
| GET | `/training/runs/{run_id}` | Auth | Get specific run details |

### Training Schemas (`api/app/schemas/training.py`)

- `TrainingStatus` — enum: pending, running, completed, failed
- `TrainingRunResponse` — id, status, started_at, completed_at, num_samples, accuracy, error_message
- `TrainingStatusResponse` — is_training, current_run

### Background Task Strategy

- Training takes ~30min-2hrs with custom training (faster than AutoML)
- Use FastAPI `BackgroundTasks` to initiate (no Celery needed at this scale)
- Training status tracked in `training_runs` table
- Admin polls `/training/status` (or web UI polls with HTMX)
- If process restarts, Vertex AI job continues independently on GCP

### Error Handling

| Failure | Handling |
|---------|----------|
| Not enough labeled images | Return 400 before starting |
| Image missing from GCS | Skip image, log warning, continue if enough remain |
| Vertex AI API error | Set TrainingRun status=failed, store error message |
| Process restart during training | Job stays pending in DB; Vertex AI job persists; admin can re-check |
| Deployment fails | status=failed; previous model stays active |
| Endpoint auto-undeploys (30 days no traffic) | Detect on next predict call, re-deploy from stored model |

### Files

**New:**
- `api/app/services/__init__.py`
- `api/app/services/ml_service.py`
- `api/app/schemas/training.py`
- `api/app/routers/training.py`

**Modified:**
- `api/app/config.py` — add ML_BACKEND, GCP_PROJECT, GCP_REGION
- `api/requirements.txt` — add google-cloud-aiplatform
- `api/app/schemas/__init__.py` — export training schemas
- `api/app/routers/__init__.py` — export training router
- `api/app/main.py` — register training router

### Tests

- Mock Vertex AI client for all tests
- Training initiation tests (start, validate data, create run)
- Status polling tests
- Completion/failure handling tests
- Mock prediction tests
- Error case tests (insufficient data, API errors)
